### PR TITLE
Create a plugin generator.

### PR DESCRIPTION
**Pull request by henders**
Wrote a generator to make creating plugins a bit easier since we are creating more and more of them.

```
$ 
$ rails generate plugin DoInterestingStuff
      create  plugins/do_interesting_stuff/lib/samson_do_interesting_stuff/samson_plugin.rb
      create  plugins/do_interesting_stuff/samson_do_interesting_stuff.gemspec
      create  plugins/do_interesting_stuff/test/test_helper.rb

To start creating your plugin, navigate to the 'samson_plugin.rb' file
in the new plugin directory above and register for the various callbacks
to hook into.

This plugin acts mostly like a Rails Engine, so it can contain controllers/
models/routes/migrations/etc...

Remember to run 'bundle install' now to get it loaded into Samson for testing!

Happy Coding!
<3 The Samson Team
$
```

This creates the following structure:
screen shot 2015-05-29 at 10 08 13

With samson_plugin.rb containing:
```ruby
module SamsonDoInterestingStuff
  class Engine < Rails::Engine
  end
end

# To add in your own UI for a hook, e.g., for the stage edit page:
# - Create app/views/samson_do_interesting_stuff/_my_fields.html.erb
# - Add this line to this file, note the lack of leading underscore and extension:
#   Samson::Hooks.view :stage_form, 'samson_do_interesting_stuff/my_fields'

#Samson::Hooks.view :stage_form, '<view path>'
#Samson::Hooks.view :deploys_header, '<view path>'
#Samson::Hooks.view :deploy_view, '<view path>'

# Possible callbacks are listed below, delete any unused ones.
#Samson::Hooks.callback :stage_clone do
  # Do stuff in here
#end

#Samson::Hooks.callback :stage_permitted_params do
  # Do stuff in here
#end

#Samson::Hooks.callback :before_deploy do
  # Do stuff in here
#end

#Samson::Hooks.callback :after_deploy do
  # Do stuff in here
#end
```

Gemspec:
```ruby
Gem::Specification.new 'samson_do_interesting_stuff', '0.0.0' do |s|
  s.summary = 'Samson DoInterestingStuff plugin'
  s.authors = ['<insert your name here>']
  s.email = ['<insert your email here>']
  s.files = Dir['{app,config,db,lib}/**/*']
end
```

/cc @zendesk/runway @zendesk/samson 

### References
 - Jira link:

### Risks
 - None

### CODE REVIEW

File: lib/generators/plugin/templates/samson_plugin.gemspec.erb [hunk 1]
```diff
@@ -0,0 +1,6 @@
+Gem::Specification.new 'samson_<%= file_name %>', '0.0.0' do |s|
+  s.summary = 'Samson <%= name %> plugin'
+  s.authors = ['<insert your name here>']
```
Comment (gabetax) [inline lib/generators/plugin/templates/samson_plugin.gemspec.erb]: Could auto-fetch these ala bundler: https://github.com/bundler/bundler/blob/master/lib/bundler/cli/gem.rb#L27-L28

File: lib/generators/plugin/templates/samson_plugin.rb.erb [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+# - Add this line to this file, note the lack of leading underscore and extension:
+#   Samson::Hooks.view :stage_form, 'samson_<%= file_name %>/my_fields'
+
+<% Samson::Hooks::VIEW_HOOKS.each do |hook| %>
+#Samson::Hooks.view :<%= hook %>, '<view path>'
+<% end %>
+
+# Possible callbacks are listed below, delete any unused ones.
+<% Samson::Hooks::EVENT_HOOKS.each do |hook| %>
+Samson::Hooks.callback :<%= hook %> do
```
Comment (gabetax) [inline lib/generators/plugin/templates/samson_plugin.rb.erb]: Maybe have all the hooks commented out so people don't accidentally leave in empty hooks?

**Comment by gabetax** [comment]
Do we want to provide any workflow opinions in regards to where these plugins get committed, i.e. to their own repo under the author, or PR'd directly into the samson trunk?

**Comment by gabetax** [comment]
:+1:

File: README.md [hunk 3]
```diff
@@ -209,6 +209,19 @@ Samson sends StatsD basic web request metrics and metrics about deploys and thre
...
+Samson now supports writing plugins to add functionality to the core app keeping the core isolated and simpler. You can
+thus add UI elements to pages that support it, and hook into events such as before and after deploys.
+
+To get started execute the following and follow the on-screen instructions:
+```
+rails generate plugin MyCoolNewPlugin
+```
+
+Also feel free to browse the existing plugins in the 'plugins' directory to get a feel. They follow the structure of a 
+Rails Engine.
```
Comment (grosser) [inline README.md]: I guess plugins/Readme.md can go then

File: lib/generators/plugin/templates/samson_plugin.rb.erb (hunk 2 above)
Comment (grosser) [inline lib/generators/plugin/templates/samson_plugin.rb.erb]: :model_defined is not used via callback

**Comment by grosser** [comment]
:+1: Looks neat / creates a clean structure with some helpers :)

**Comment by henders** [comment]
Cool, thanks for the feedback @gabetax and @grosser 

I've updated it to auto-populate the git username/email in the gemspec, and removed the unused :model_defined callback, and commented out the hooks.

**Comment by grosser** [comment]
:+1: looks good!

**Pull request merged**